sycl/doc/extensions/proposed/sycl_ext_oneapi_graph.asciidoc

```diff
@@ -0,0 +1,1637 @@
+
+When one thread ends recording on a queue while another
+thread is submitting work, which kernels will be part of the subsequent
+graph is undefined. If user code enforces a total order on the queue
+events, then the behavior is well-defined, and will match the observable
+total order.
+
+The returned value from the `queue::ext_oneapi_get_state()` should be
+considered immediately stale in multi-threaded usage, as another thread could
+have preemptively changed the state of the queue.
+
+=== Exception Safety
+
+In addition to the destruction semantics provided by the SYCL
+{crs}[common reference semantics], when the last copy of a modifiable
+`command_graph` is destroyed recording is ended on any queues that are recording
+to that graph, equivalent to `+this->end_recording()+`.
+
+As a result, users don't need to manually wrap queue recording code in a
+`try` / `catch` block to reset the state of recording queues on an exception
+back to the executing state. Instead, an uncaught exception destroying the
+modifiable graph will perform this action, useful in RAII pattern usage.
+
+=== Storage Lifetimes [[storage-lifetimes]]
+
+The lifetime of any buffer recorded as part of a submission
+to a command graph will be extended in keeping with the common reference
+semantics and buffer synchronization rules in the SYCL specification. It will be
+extended either for the lifetime of the graph (including both modifiable graphs
+and the executable graphs created from them) or until the buffer is no longer
+required by the graph (such as after being replaced through executable graph update).
+
+If a buffer created with a host data pointer is recorded as part of a submission to
+a command graph, the lifetime of that host data will also be extended by taking a
+copy of that data inside the buffer. This copy will occur during the call to
+`queue::submit()`. Users can opt-out of this behavior by passing the property
+`graph::no_host_copy` to `command_graph::begin_recording()` or the constructor
+of the `command_graph` the commands will be recorded to. Passing the property to
+`begin_recording()` will prevent host copies only for commands recorded before
+`end_recording()` is called for a given queue. Passing the property to the
+`command_graph` constructor will prevent host copies for all commands recorded to
+the graph.
+
+This behavior was introduced to address the case where host data used to create
+a buffer is destroyed before the copy to device has happened. For example:
+[source,c++]
+----
+void foo(queue q /* queue in recording mode */ ) {
+  float data[NUM];
+  buffer buf{data, range{NUM}};
+  q.submit([&](handler &cgh) {
+    accessor acc{buf, cgh, read_only};
+    cgh.single_task([] {
+       // use "acc"
+    });
+  });
+  // "data" goes out of scope
+}
+----
+
+The copy will not occur if the buffer is accessed inside the submitted command
+group function with an accessor with `access_mode::write` or the `no_init`
+property.
```

Comment:
This final sentence still does not make sense to me.  If the access mode is `write`, then the buffer has copy-back semantics, and we say later that it is an error to record a buffer with copy-back semantics.  The same is true if the access mode is `read_write`.  The `no_init` property can only be used for accessors that are `write` or `read_write`.  Therefore, this sentence seems kind of pointless because these conditions all lead to errors anyway.

Reply:
I'm not sure I understand your comment. Doesn't an access mode of `write` just imply that the kernel will write data to the buffer/device memory? If copy-back is disabled on the buffer as we require then a `write` accessor doesn't imply any additional copy-back. I believe it is true that if no kernels write data to a buffer that has copy-back enabled then no copy-back will occur because there is no change in the data. But that seems moot in this case because we require disabling the copy-back mechanism altogether.

I also don't think this idea of copying the host data inside the buffer in these cases has much to do with copy-back. In this case we're purely worried about the user deleting the buffer in their code before command-graph execution (where we have to extend the lifetime) and then thinking it's safe to delete the host data which has yet to be copied to the buffer (which happens on first execution of the graph).

Reply:
Looks like this sentence is controversial. 

I would suggest: Let's remove it for now, a user who wants to enforce the described no-copy behavior can still use the `graph::no_host_copy` property. If we think it really adds value to be more restrictive here about host copy we can update in a future version.

Reply:
Thanks for the explanation @Bensuo.  Feel free to either keep the sentence or remove it as you see fit.  If you keep it, though, I think it would help to mention the part about disabling copy-back.  For example, something like:

> The implementation guarantees that the host memory will not be copied internally if all the commands accessing this buffer use `access_mode::write` or the `no_init` property because the host memory is not needed in these cases.  Note, however, that these cases require the application to disable copy-back as described in \<\<buffer-limitations\>\>.

Reply:
Thanks for the suggestion! This has been added as part of https://github.com/reble/llvm/pull/231.